api/v1alpha7/types.go

```diff
@@ -163,20 +163,83 @@ type RootVolume struct {
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
 }
 
-type AdditionalBlockDevice struct {
-	// Name of the Cinder volume in the context of a machine.
-	// It will be combined with the machine name to make the actual volume name.
-	Name string `json:"name"`
-	// Size is the size in GB of the volume.
-	Size int `json:"diskSize"`
-	// VolumeType is the volume type of the volume.
-	// If omitted, the default type will be used.
-	VolumeType string `json:"volumeType,omitempty"`
-	// AvailabilityZone is the volume availability zone to create the volume in.
+// blockDeviceStorage is the storage type of a block device to create and
+// contains additional storage options.
+// +kubebuilder:validation:XValidation:rule="self.type == 'Local' ?  !has(self.volume) : has(self.volume)",message="volume is forbidden for type 'Local'"
```

Comment:
@JoelSpeed the only rule that I think we need. Since `volume` is optional, I don't think we should have any other CEL here.